concurrent/src/main/java/io/airlift/concurrent/DynamicSizeBoundQueue.java

```diff
@@ -0,0 +1,337 @@
+import jakarta.annotation.Nullable;
+
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.ToLongFunction;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+import static java.util.Objects.requireNonNullElseGet;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * Size constrained queue that utilizes a dynamic element size function. To prevent
+ * starvation for adding large elements, the queue will only block new elements if
+ * the total size has already been reached or exceeded. This means in the normal
+ * case, the queue should be no larger than the max size plus the size of one element.
+ * Callers also have the additional option to force insert further elements without
+ * regard for size constraints. In the current implementation, elements are required
+ * to have positive sizes (they cannot have zero size). This implementation is designed
+ * to closely mirror the method signatures of {@link java.util.concurrent.BlockingQueue}.
+ */
+@ThreadSafe
+public class DynamicSizeBoundQueue<T>
+{
+    private final AtomicLong size = new AtomicLong();
+    private final Queue<ElementAndSize<T>> queue = new ConcurrentLinkedQueue<>();
+    private final AtomicReference<SettableFuture<Void>> enqueueFuture = new AtomicReference<>();
+    private final AtomicReference<SettableFuture<Void>> dequeueFuture = new AtomicReference<>();
+
+    private final long maxSize;
+    private final ToLongFunction<T> elementSizeFunction;
+    private final Ticker ticker;
+
+    public DynamicSizeBoundQueue(long maxSize, ToLongFunction<T> elementSizeFunction)
+    {
+        this(maxSize, elementSizeFunction, Ticker.systemTicker());
+    }
+
+    public DynamicSizeBoundQueue(long maxSize, ToLongFunction<T> elementSizeFunction, Ticker ticker)
+    {
+        checkArgument(maxSize > 0, "maxSize must be positive");
+        this.maxSize = maxSize;
+        this.elementSizeFunction = requireNonNull(elementSizeFunction, "elementSizeFunction is null");
+        this.ticker = requireNonNull(ticker, "ticker is null");
+    }
+
+    public long getMaxSize()
+    {
+        return maxSize;
+    }
+
+    /**
+     * Gets the current size of the queue. The size is guaranteed to be no larger than max size plus
+     * the size of one element if {@link DynamicSizeBoundQueue#forcePut(Object)} is not used.
+     */
+    public long getSize()
+    {
+        return size.get();
+    }
+
+    public boolean offer(T element)
+    {
+        long elementSize = elementSizeFunction.applyAsLong(element);
```

Comment:
I suggest adding a function to do this computation, so you can also verify the element is not null, and check that the result is not negative